nomenklatura/index/duckdb_index.py

```diff
@@ -0,0 +1,217 @@
+        """Yields tuples of (field_name, token)"""
+        query = """
+            SELECT field, token, count(*) as frequency
+            FROM entries
+            GROUP BY field, token
+        """
+        token_counts_rel = self.con.sql(query)
+        filter_query = "SELECT * from token_counts_rel where frequency > 100"
+        common_tokens_rel = self.con.sql(filter_query)
+        tokens: Set[Tuple[str, str]] = set()
+        for field_name, token, freq in common_tokens_rel.fetchall():
+            tokens.add((field_name, token))
+        return tokens
+
+    def id_grouped_mentions(
+        self,
+    ) -> Generator[Tuple[str, str, int, List[Tuple[str, int]]], None, None]:
+        """
+        Yields tuples of (field_name, entity_id, field_len, [(token, mention_count)])
+        """
+        common_tokens = self.common_tokens()
+        mentions_gen = self.mentions()
+        mention_row = None
+        # Read all field lengths into memory because the concurrent iteration
+        # sees to be exiting the outer loop early and giving partial results.
+        for field_name, id, field_len in list(self.field_lengths()):
+            mentions = []
+            try:
+                if mention_row is None:  # first iteration
+                    mention_row = next(mentions_gen)
+                    mention_field_name, mention_id, token, mention_count = mention_row
+
+                while mention_field_name == field_name and mention_id == id:
+                    if (mention_field_name, token) not in common_tokens:
+                        mentions.append((token, mention_count))
+
+                    mention_row = next(mentions_gen)
+                    mention_field_name, mention_id, token, mention_count = mention_row
+
+                yield field_name, id, field_len, mentions
+
+            except StopIteration:
+                yield field_name, id, field_len, mentions
+                break
+
+    def calculate_frequencies(self) -> None:
+        csv_path = self.path / "frequencies.csv"
+        with open(csv_path, "w") as fh:
+            writer = csv.writer(fh)
+            writer.writerow(["field", "id", "token", "frequency"])
+
+            for field_name, id, field_len, mentions in self.id_grouped_mentions():
+                for token, freq in mentions:
+                    writer.writerow([field_name, id, token, freq / field_len])
+
+        log.info(f"Loading frequencies data... ({csv_path})")
+        self.con.execute(
+            "CREATE TABLE frequencies (field TEXT, id TEXT, token TEXT, frequency FLOAT)"
+        )
+        self.con.execute(f"COPY frequencies from '{csv_path}'")
+        log.info("Frequencies are loaded")
+
+    def frequencies(
+        self,
+    ) -> Generator[Tuple[str, str, List[Tuple[Identifier, float]]], None, None]:
+        query = """
+            SELECT field, token, id, frequency
+            FROM frequencies
+            ORDER by field, token
```

Comment:
This is a bit weird, but: is there a world where we want to do `ORDER BY frequency DESC OFFSET 1000` (or OFFSET 1%) here to basically take out stopword tokens?